lib/core/connection/utils.js

```diff
@@ -16,6 +19,14 @@ function retrieveBSON() {
   BSON.native = false;
 
   const optionalBSON = require_optional('bson-ext');
+  const bsonExtVersion = parsePackageVersion(
+    require_optional('bson-ext/package.json') || { version: '0.0.0' }
+  );
+  if (bsonExtVersion.major >= 4) {
+    emitWarningOnce(
+      'bson-ext version 4 and above does not work with the 3.x version of the mongodb driver'
+    );
```

Comment:
Since I added the version parsing logic I figured we could add a check here to warn folks, we could also throw here, but the driver will crash else where anyway "bson is not a constructor" is likely the first error to be hit.

Reply:
wouldn't it be more helpful to throw here with the appropriate error instead of letting it fail in a weird way down the line? also, why not put the code doing the check in the `if (optionalBSON)` block?

Reply:
Changed it, and I tested this, it throws at the moment of `require('mongodb')` so it'll stop scripts right up front.